Implement transition event and infrastructure

KiChjang opened the pull request:
Fixes #10245.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/13056)

<!-- Reviewable:end -->

KiChjang committed:
Implement TransitionEvent

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/data.rs, components/style/traversal.rs, components/style/matching.rs, components/style/Cargo.toml, components/style/lib.rs, components/style/dom.rs, components/style/animation.rs, components/style/context.rs, components/style/parallel.rs
 * @kichjang: components/script/dom/webidls/EventHandler.webidl, components/script/dom/webidls/TransitionEvent.webidl, components/script/dom/mod.rs, components/script_traits/lib.rs, components/script_traits/lib.rs, components/script/lib.rs, components/script/dom/macros.rs, components/script/Cargo.toml, components/script/script_thread.rs, components/script/layout_wrapper.rs, components/script/dom/transitionevent.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify style, layout, and script code, but no tests are modified. Please consider adding a test!

KiChjang commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 80ead58 with merge e2becc6...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/2724)

KiChjang commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 5adb03c with merge 6e54ed7...

jdm commented:
r? @nox 
I'm going to be away all next week and won't be able to review this promptly.

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/2728)

jdm commented:
We should modify the test mentioned in https://github.com/servo/servo/issues/10473#issuecomment-207536841 and reenable it.

KiChjang commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit aa66059 with merge 1168383...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/2732)

KiChjang commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit b4d6459 with merge 3a05450...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2896)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13098) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13106) made this pull request unmergeable. Please resolve the merge conflicts.

nox commented:
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 3 of 3 files at r1, 22 of 22 files at r2, 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[components/script/dom/transitionevent.rs, line 21 \[r1\]](https://reviewable.io:443/reviews/servo/servo/13056#-KQUtwgOCf893CP1_a06:-KQUtwgOCf893CP1_a07:bgl5pa4) ([raw file](https://github.com/servo/servo/blob/9fbf3f0f1c56df449ae5dc9f9eba7de132ce8c8d/components/script/dom/transitionevent.rs#L21)):*
> ```Rust
> pub struct TransitionEvent {
>     event: Event,
>     property_name: DOMString,
> ```

This should probably be an `Atom`.

---

*[components/script/dom/transitionevent.rs, line 23 \[r1\]](https://reviewable.io:443/reviews/servo/servo/13056#-KQUtz1Ip1J1x5lv6CQm:-KQUtz1Ip1J1x5lv6CQn:bgl5pa4) ([raw file](https://github.com/servo/servo/blob/9fbf3f0f1c56df449ae5dc9f9eba7de132ce8c8d/components/script/dom/transitionevent.rs#L23)):*
> ```Rust
>     property_name: DOMString,
>     elapsed_time: Finite<f32>,
>     pseudo_element: DOMString,
> ```

This should probably be an `Atom`.

---

*[components/script/dom/webidls/EventHandler.webidl, line 92 \[r2\]](https://reviewable.io:443/reviews/servo/servo/13056#-KQUvVA3ChliPaU3zPzq:-KQUvVA3ChliPaU3zPzr:b637t68) ([raw file](https://github.com/servo/servo/blob/2318a7baa75d9485c1d91300a438fb62e234e9be/components/script/dom/webidls/EventHandler.webidl#L92)):*
> ```WebIDL
> 
> // https://drafts.csswg.org/css-transitions/#interface-globaleventhandlers-idl
>            attribute EventHandler ontransitionend;
> ```

Please do a separate partial interface like in the spec you linked.

---

*[components/style/animation.rs, line 723 \[r3\]](https://reviewable.io:443/reviews/servo/servo/13056#-KQUwEqEZBpK3m49sfpf:-KQUwEqEZBpK3m49sfpg:b1e3mfv) ([raw file](https://github.com/servo/servo/blob/486857a373291402ec1b4fc27478d94991437229/components/style/animation.rs#L723)):*
> ```Rust
> 
>     had_animations_to_expire
> }
> ```

Please dedup that code.

---

*[components/style/Cargo.toml, line 34 \[r2\]](https://reviewable.io:443/reviews/servo/servo/13056#-KQUuU-vOVSbSZ9dOr9u:-KQUuU-vOVSbSZ9dOr9v:bj0neyz) ([raw file](https://github.com/servo/servo/blob/2318a7baa75d9485c1d91300a438fb62e234e9be/components/style/Cargo.toml#L34)):*
> ```TOML
> heapsize = {version = "0.3.0", optional = true}
> heapsize_plugin = {version = "0.1.2", optional = true}
> ipc-channel = "0.5"
> ```

How does geckolib cope with that?

---

*[components/style/context.rs, line 79 \[r2\]](https://reviewable.io:443/reviews/servo/servo/13056#-KQUucXwXAfSzMEM0sxT:-KQUucXwXAfSzMEM0sxU:b7ityv9) ([raw file](https://github.com/servo/servo/blob/2318a7baa75d9485c1d91300a438fb62e234e9be/components/style/context.rs#L79)):*
> ```Rust
> 
>     /// A channel to communicate with the script thread about transition events.
>     pub script_chan: IpcSender<ConstellationControlMsg>,
> ```

I'm a bit wary of this being a `ConstellationControlMsg` sender when the only things we want to send are transitionend notifications.

---

*[components/style/context.rs, line 26 \[r3\]](https://reviewable.io:443/reviews/servo/servo/13056#-KQUwOnJLUDTXnqOxkC3:-KQUwOnJLUDTXnqOxkC4:bahqaxv) ([raw file](https://github.com/servo/servo/blob/486857a373291402ec1b4fc27478d94991437229/components/style/context.rs#L26)):*
> ```Rust
> pub struct LocalStyleContextCreationInfo {
>     new_animations_sender: Sender<Animation>,
>     script_chan: IpcSender<ConstellationControlMsg>,
> ```

Can't you just put `#[cfg(feature = "servo")]` on the field?

---

*[components/style/context.rs, line 96 \[r3\]](https://reviewable.io:443/reviews/servo/servo/13056#-KQUwW2bxM6Ab11yKPjz:-KQUwW2bxM6Ab11yKPk-:bxn9ca2) ([raw file](https://github.com/servo/servo/blob/486857a373291402ec1b4fc27478d94991437229/components/style/context.rs#L96)):*
> ```Rust
> 
>     /// A channel to communicate with the script thread about transition events.
>     pub script_chan: IpcSender<ConstellationControlMsg>,
> ```

Pretty sure you can just `cfg`-gate the field.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/13056#-:-KQUwc4kaK6hPbujN6iX:bo47p9t)*
<!-- Sent from Reviewable.io -->

KiChjang commented:
Currently this doesn't work yet, there are tests that it should be passing but isn't. I need to figure out why transition events aren't being fired...

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13110) made this pull request unmergeable. Please resolve the merge conflicts.

KiChjang committed:
Emit TransitionEnd events in the layout thread and process it in the script thread

KiChjang commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 0dd7048 with merge 993bc44...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2848)

KiChjang commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 26a0970 with merge f46d70f...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/3013)

KiChjang commented:
I think this is ready for review again.

@bors-servo try

bors-servo commented:
:hourglass: Trying commit 660d7a4 with merge 784c94d...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/2852)

emilio commented:
There's a typo in the commit message ("proces").

---

So I've seen a bunch of it and, while this should work fine, I'm kind of wary about the usage of `UnsafeNode` moving into script, since it's supposed to be a layout-only stuff.

I'd like to find another less unsafe way of doing that, but I can't think of anything easy off-hand. I'd like @pcwalton to sign-off this though.

---

Review status: 2 of 24 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[components/script/script_thread.rs, line 1542 \[r5\]](https://reviewable.io:443/reviews/servo/servo/13056#-KQnZqijJ-qhz8DiEpEu:-KQnZqijJ-qhz8DiEpEv:b7pal4x) ([raw file](https://github.com/servo/servo/blob/660d7a4bee480991170a6c0028e93e29237c581e/components/script/script_thread.rs#L1542)):*
> ```Rust
>     fn handle_transition_event(&self, unsafe_node: UnsafeNode, name: Atom, duration: f64) {
>         let node = unsafe { ServoLayoutNode::from_unsafe(&unsafe_node) };
>         let node = unsafe { node.get_jsmanaged().get_for_script() };
> ```

This roundtrip should be completely unnecessary. It's probably a better approach to implement `Node::from_unsafe` instead, and use that from `ServoLayoutNode`, or something. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/13056#-:-KQnZdUL1F7FImlDUivM:b3fxnj9)*
<!-- Sent from Reviewable.io -->

KiChjang commented:
@bors-servo try

KiChjang commented:
@bors-servo try- retry clean try

KiChjang commented:
@bors-servo try- clean

KiChjang commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 3c5b280 with merge b75bb09...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/2857)

nox commented:
```
/home/servo/buildbot/slave/linux-dev/build/target/geckolib/debug/build/style-a3eeb74df56c5f99/out/properties.rs:35008:71: 35008:92 error: no rules expected the token `"background-position"`
/home/servo/buildbot/slave/linux-dev/build/target/geckolib/debug/build/style-a3eeb74df56c5f99/out/properties.rs:35008                     AnimatedProperty::BackgroundPosition(..) => atom!("background-position"),
```

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13180) made this pull request unmergeable. Please resolve the merge conflicts.

KiChjang commented:
The reason for the test failures here is because gecko doesn't recognize the `background-position` atom. I've chatted with @emilio about this and it looks like we can either revert to passing a String, or patch gecko so that it properly recognizes `background-position`.

KiChjang commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit ddf34c3 with merge 8761270...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/2959)

KiChjang commented:
Okay, this is going nowhere, I'm not attempting to use atoms for now.

@bors-servo try

bors-servo commented:
:hourglass: Trying commit 7643971 with merge 1f1be04...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2959)

ConnorGBrewster commented on components/script/dom/transitionevent.rs:
```diff
@@ -0,0 +1,74 @@
+    #[allow(unrooted_must_root)]
+    pub fn new_inherited(type_: Atom, init: &TransitionEventInit) -> TransitionEvent {
+        let event = Event::new_inherited();
+        event.init_event(type_, init.parent.bubbles, init.parent.cancelable);
```
nit: In the other event structs the `new_inherited` usually doesn't handle calling `init_event`, this is typically done in `fn new()`. This would avoid `#[allow(unrooted_must_root)]`

see https://dxr.mozilla.org/servo/source/components/script/dom/closeevent.rs#26-33,41-58

KiChjang commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 16f8ae2 with merge 3108c69...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/3132)

KiChjang commented:
@jdm/@Ms2ger I'm going to expose the TransitionEvent object on the global.

KiChjang commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 418052f with merge b4f554f...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/3134)

KiChjang commented:
r? @mbrubeck

mbrubeck commented:
Reviewed 11 of 24 files at r4, 3 of 22 files at r6, 17 of 20 files at r7, 3 of 5 files at r8.
Review status: 21 of 24 files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/13056)*
<!-- Sent from Reviewable.io -->

pcwalton commented on components/script/dom/bindings/js.rs:
```diff
@@ -437,6 +437,13 @@ impl<T: Reflectable> LayoutJS<T> {
+
+    /// Returns a reference to the interior of this JS object. This method is
+    /// safe to call because it originates from the layout thread, and it cannot
+    /// mutate DOM nodes.
```
Are you sure? How do you enforce that only the layout thread is calling this?

KiChjang replied:
On the contrary, this should always be only called by the script thread, hence the `for_script` prefix. But you're right that I need an assertion here.

pcwalton commented on components/script/script_thread.rs:
```diff
@@ -1526,6 +1535,33 @@ impl ScriptThread {
+    /// Handles firing of transition events.
+    #[allow(unsafe_code)]
+    fn handle_transition_event(&self, unsafe_node: UnsafeNode, name: String, duration: f64) {
+        let node = unsafe { ServoLayoutNode::from_unsafe(&unsafe_node) };
```
How do you ensure that the DOM is locked at this point?

emilio replied:
Isn't this the script thread? Seeing ServoLayoutNode here feels weird anyway, I agree.

pcwalton reviewed (changes requested):
This looks racy to me. I think you should store the "this transition has an event attached" flag ahead of time in the layout node and send a message back to script with the address of the node so that it can do the right thing.

emilio commented on components/script/script_thread.rs:
```diff
@@ -1526,6 +1535,33 @@ impl ScriptThread {
+            },
+            propertyName: DOMString::from(name),
+            elapsedTime: Finite::new(duration as f32).unwrap(),
+            pseudoElement: DOMString::new()
```
Do we handle pseudo-elements at all?

It seems to me that we we might be animating the node if the pseudo-element style is animated instead, though that seems a more general issue not related to the DOM bits only (I'm on an airport right now and can't check easily).

If you don't fix it as part of this issue (which would be awesome), please open an issue about it and CC me, then reference it from here.

KiChjang replied:
I wanted to, but I'm not entirely sure how to. How can you extract information about whether a transition is being called upon a pseudo-element instead of an element if you have the element node itself?

emilio replied:
You can pass it down to `cascade_node_pseudo_element`, instead of the bool `should_animate_properties`.

mbrubeck commented:
I think this is ready with a couple of nits (below) addressed:

---

Reviewed 1 of 3 files at r1, 1 of 5 files at r8, 12 of 12 files at r9.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[components/script/script_thread.rs, line 1557 at r8](https://reviewable.io:443/reviews/servo/servo/13056#-KQ6Z9BGg28IKRYyBWFJ-r8-1557:-KToiDfVZRqks39HnMCi:byboirt) ([raw file](https://github.com/servo/servo/blob/bae97662dc3929d0f98bf398058098e93e4744d0/components/script/script_thread.rs#L1557)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary>
> You can pass it down to `cascade_node_pseudo_element`, instead of the bool `should_animate_properties`.
</details>
Could you add a FIXME comment here (assuming this is not fixed yet)?

---

*[components/script/dom/transitionevent.rs, line 23 at r1](https://reviewable.io:443/reviews/servo/servo/13056#-KQUtz1Ip1J1x5lv6CQm:-KToic8d0measSb6nHKg:b747486) ([raw file](https://github.com/servo/servo/blob/9fbf3f0f1c56df449ae5dc9f9eba7de132ce8c8d/components/script/dom/transitionevent.rs#L23)):*
<details><summary><i>Previously, nox (Anthony Ramine) wrote…</i></summary>
> This should probably be an `Atom`.
</details>
Can this be changed to an `Atom`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/13056#-:-KToirJEX0vGVvh6UCP8:b-qfvliy)*
<!-- Sent from Reviewable.io -->

KiChjang commented:
@mbrubeck I tried changing the Strings into Atoms, but this involves patching stylo as well, since some keywords are not made into proper atoms in stylo's string_cache.

mbrubeck commented:
> I tried changing the Strings into Atoms, but this involves patching stylo as well, since some keywords are not made into proper atoms in stylo's string_cache.

Okay, it's fine to leave this as a follow-up.

KiChjang commented:
Comments addressed.

mbrubeck commented:
@bors-servo r+

---

Reviewed 10 of 10 files at r10.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/13056#-:-KTpqtFaIAT0csKtUASB:bkg9nno)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:pushpin: Commit 260e24d has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck 260e24dd9d41f16f1ad05f2589b71fd926a68f96 -->

bors-servo commented:
:hourglass: Testing commit 260e24d with merge a6f33e3...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/172)

mbrubeck commented:
```
  ▶ Unexpected subtest result in /_mozilla/mozilla/interfaces.worker:
  │ FAIL [expected PASS] Untitled
  │   → assert_true: TransitionEvent should be defined on the global scope expected true got false
```

KiChjang commented:
Fixed the interface error (hopefully). @bors-servo r=mbrubeck
* #11266

bors-servo commented:
:pushpin: Commit 0270858 has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck 0270858cd5cb67614e9ef25d9d0c94e4cf88ab2f -->

jdm commented:
@bors-servo: r-
Travis and appveyor disagree. Also, please don't expose TransitionEvent to workers, only Window.

jdm commented:
Sorry, didn't realize CI failures were preexisting.

jdm commented:
@bors-servo: r=mbrubeck

bors-servo commented:
:pushpin: Commit 0270858 has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck 0270858cd5cb67614e9ef25d9d0c94e4cf88ab2f -->

bors-servo commented:
:hourglass: Testing commit 0270858 with merge d210549...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/175)

KiChjang commented:
This still shows up:
```
Tests with unexpected results:
  ▶ Unexpected subtest result in /_mozilla/mozilla/interfaces.worker:
  │ FAIL [expected PASS] Untitled
  │   → assert_true: TransitionEvent should be defined on the global scope expected true got false
  │ 
  │ test_interfaces/<@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:83:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1401:20
  │ test@http://web-platform.test:8000/resources/testharness.js:501:9
  │ test_interfaces@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:2:3
  └ @http://web-platform.test:8000/_mozilla/mozilla/interfaces.worker.js:9:1
```
I modified `interfaces.workers.js` before and added TransitionEvent, but it didn't seem to solve it.

jdm commented on tests/wpt/mozilla/tests/mozilla/interfaces.js:
```diff
@@ -40,6 +40,7 @@ function test_interfaces(interfaceNamesInGlobalScope) {
       "String",
       "Symbol",
       "SyntaxError",
+      "TransitionEvent",
```
This is incorrect. It is not an EcmaScript global property.

jdm commented on tests/wpt/mozilla/tests/mozilla/interfaces.worker.js:
```diff
@@ -101,6 +101,7 @@ test_interfaces([
   "Touch",
   "TouchEvent",
   "TouchList",
+  "TransitionEvent",
```
There should be no need to expose this interface to workers, so let's remove this.

jdm commented on components/script/dom/webidls/TransitionEvent.webidl:
```diff
@@ -0,0 +1,20 @@
+ * https://dom.spec.whatwg.org/#event
+ */
+
+[Constructor(DOMString type, optional TransitionEventInit transitionEventInitDict)]
```
Add `Exposed=Window` as well.

KiChjang commented:
'kay, let's try this again. @bors-servo r=mbrubeck

bors-servo commented:
:pushpin: Commit 668163e has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck 668163ec5c4091806b155ef14d3b3522cb4697cd -->

bors-servo commented:
:hourglass: Testing commit 668163e with merge 16570b7...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3470), [arm64](http://build.servo.org/builders/arm64/builds/3482), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3555), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/176), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/177), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3547), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3526), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/228), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/228), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1435)

bors-servo commented:
:eyes: Test was successful, but fast-forwarding failed: 422 Update is not a fast forward

Manishearth commented:
@bors-servo retry
 - infra bork

bors-servo commented:
:hourglass: Testing commit 668163e with merge 5a14d09...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/230)

Manishearth commented:
@bors-servo retry

 - #13662

bors-servo commented:
:hourglass: Testing commit 668163e with merge cd2f950...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3477), [arm64](http://build.servo.org/builders/arm64/builds/3489), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3561), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/183), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/184), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3554), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3533), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/235), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/235), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1442)
